keras_nlp/models/gpt2/gpt2_backbone.py

```diff
@@ -170,6 +174,116 @@ def __init__(
         self.intermediate_dim = intermediate_dim
         self.dropout = dropout
         self.max_sequence_length = max_sequence_length
+        self.transformer_layers = []
```

Comment:
I think rather than add a bunch of attributes, we should just move the `self.get_layer("name")` calls into the call_with_cache method.

Otherwise we are playing a dangerous game, as adding a layer that is not a toplevel of the functional model to self will break all our checkpoints!

Reply:
I was doing this way to avoid the overhead of massive `get_layer` calls, and for some readability. Also I am unsure how the layer is recognized, by doing this way we can still load the old checkpoints.

But anyway, it is fine to me to move these `get_layer` into `call_with_cache`, which should not significantly affect the performance.

Reply:
> Also I am unsure how the layer is recognized, by doing this way we can still load the old checkpoints.

It is a bit complicated, we can talk offline. Basically you won't see any issues `backbone.save_weights`/`backbone.load_weights`, but your changes (as currently stand on the PR) would change `causal_lm.save_weights`/`causal_lm.load_weights` because the causal lm will think it has all these extra layers!

Another way to see this is just printing `causal_lm.summary()`. The causal lm will print a backbone with `124m` parameters, but then duplicate all the individual layers in the summary again.

Happy to talk through more, but the short version is we can either do `get_layer("name")` calls, or the setattr tricks Francois mentioned below. `get_layer("name")` call straight in `call_with_cache` seems the simplest to get this landed.

Reply:
thx for the explanation, i will refactor the code.